File: ios/Wrappers/CreateGroupParamsWrapper.swift
import Foundation
import XMTP

struct CreateGroupParamsWrapper {
	let groupName: String
	let groupImageUrl: String
	let groupDescription: String
	let disappearingMessageSettings: DisappearingMessageSettings?
	let appData: String?

	static func createGroupParamsFromJson(_ authParams: String)
		-> CreateGroupParamsWrapper
	{
		let data = authParams.data(using: .utf8) ?? Data()
		let jsonOptions =
			(try? JSONSerialization.jsonObject(with: data, options: []))
				as? [String: Any] ?? [:]

		var settings: DisappearingMessageSettings? = nil

		// Only create DisappearingMessageSettings if both values are provided
		if let disappearStartingAtNs = jsonOptions["disappearStartingAtNs"] as? Int64,
		   let retentionDurationInNs = jsonOptions["retentionDurationInNs"] as? Int64
		{
			settings = DisappearingMessageSettings(
				disappearStartingAtNs: disappearStartingAtNs,
				retentionDurationInNs: retentionDurationInNs
			)
		}

		let groupName = jsonOptions["name"] as? String ?? ""
		let groupImageUrl = jsonOptions["imageUrl"] as? String ?? ""
		let groupDescription = jsonOptions["description"] as? String ?? ""
		let appData = jsonOptions["appData"] as? String

		return CreateGroupParamsWrapper(
			groupName: groupName,
			groupImageUrl: groupImageUrl,
			groupDescription: groupDescription,
			disappearingMessageSettings: settings,
			appData: appData
		)
	}
}

File: android/src/main/java/expo/modules/xmtpreactnativesdk/wrappers/CreateGroupParamsWrapper.kt
package expo.modules.xmtpreactnativesdk.wrappers

import com.google.gson.JsonParser
import org.xmtp.android.library.libxmtp.DisappearingMessageSettings

class CreateGroupParamsWrapper(
    val groupName: String,
    val groupImageUrl: String,
    val groupDescription: String,
    val disappearingMessageSettings: DisappearingMessageSettings?,
    val appData: String,
) {
    companion object {
        fun createGroupParamsFromJson(authParams: String): CreateGroupParamsWrapper {
            val jsonOptions = JsonParser.parseString(authParams).asJsonObject

            // Only create DisappearingMessageSettings if both values are provided
            val settings = if (jsonOptions.has("disappearStartingAtNs") && jsonOptions.has("retentionDurationInNs")) {
                DisappearingMessageSettings(
                    jsonOptions.get("disappearStartingAtNs").asLong,
                    jsonOptions.get("retentionDurationInNs").asLong
                )
            } else {
                null
            }

            return CreateGroupParamsWrapper(
                if (jsonOptions.has("name")) jsonOptions.get("name").asString else "",
                if (jsonOptions.has("imageUrl")) jsonOptions.get("imageUrl").asString else "",
                if (jsonOptions.has("description")) jsonOptions.get("description").asString else "",
                settings,
                if (jsonOptions.has("appData")) jsonOptions.get("appData").asString else "",
            )
        }
    }
}
